### PR TITLE
8262981: Create implementation for NSAccessibilitySlider protocol

### PR DESCRIPTION
**Pull request by pankaj-bansal**
Create implementation of NSAccessibilitySlider protocol https://developer.apple.com/documentation/appkit/nsaccessibilityslider

The implementation has the function performIncrement and performDecrement to increase/decrease the value of slider using the VoiceOver. To implement this functionality, I could think of  following two ways. I have chosen the first one here though it is more intrusive.
1. Make the AccessibleJSlider class implement the AccessibleAction interface. This makes AccessibleJSlider consistent with the AccessibleJSpinner class. It is more clear and the logic to increase/decrease Slider value can be changed easily. But this changes AccessibleJSlider class and will need a CSR.
2. Get the current Accessible Value from the component and  just increment/decrement it in SliderAccessibility.m  class itself and then set the current accessible value fro there only. This will not need any changes in AccessibleJSlider class, but this does not look correct way and I have not used this.

The changes can be easily tested by using a JSlider example, like the following example. VO should announce the correct the slider values. To change the slider values, use ctrl+opt+shift+down key to start interacting with the slider, then use ctrl+opt+up/down arrow keys to increment/decrement slider values.

      import javax.swing.JFrame;
      import javax.swing.JSlider;
      
      public class JSliderDemo{
          public static void main(String[] args) throws Exception {
              JFrame jframe = new JFrame("JSlider Demo");
              jframe.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
              jframe.add(new JSlider());
      
              jframe.setLocationRelativeTo(null);
              jframe.pack();
              jframe.setVisible(true);
          }
      }

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8262981](https://bugs.openjdk.java.net/browse/JDK-8262981): Create implementation for NSAccessibilitySlider protocol

### Reviewers
 * [Alexander Zuev](https://openjdk.java.net/census#kizune) (@azuev-java - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/2874/head:pull/2874` \
`$ git checkout pull/2874`

Update a local copy of the PR: \
`$ git checkout pull/2874` \
`$ git pull https://git.openjdk.java.net/jdk pull/2874/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 2874`

View PR using the GUI difftool: \
`$ git pr show -t 2874`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/2874.diff">https://git.openjdk.java.net/jdk/pull/2874.diff</a>

</details>

### CODE REVIEW

File: src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m [hunk 1]
```diff
@@ -56,6 +56,7 @@ + (void) initializeRolesMap {
     [rolesMap setObject:@"StaticTextAccessibility" forKey:@"label"];
     [rolesMap setObject:@"RadiobuttonAccessibility" forKey:@"radiobutton"];
     [rolesMap setObject:@"CheckboxAccessibility" forKey:@"checkbox"];
+    [rolesMap setObject:@"SliderAccessibility" forKey:@"slider"];
```
Comment (azuev-java) [inline src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m:59]: Shouldn't you also increase the rolesMap initial capacity?

**Comment by azuev-java** [comment]
> If you need a new functionality and CSR I wonder how it currently works on Windows and macOS, or it does not work?

Haven't tested this functionality on Windows, on macOS with current JDK it does not work at all. After entering the JSlider interaction mode VoiceOver does not announce the a11y shortcuts and the attempt to invoke said shortcuts do nothing.

**Comment by pankaj-bansal** [comment]
> If you need a new functionality and CSR I wonder how it currently works on Windows and macOS, or it does not work?

File: src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m (hunk 1 above)
Comment (pankaj-bansal) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m:59]: > Shouldn't you also increase the rolesMap initial capacity?

ok, I will do that in next commit.

**Comment by mrserb** [comment]
If it does not work on Windows then I suggest checking that the new functionality requested via CSR will help it as well.

**Comment by azuev-java** [comment]
> If these changes are too intrusive, I can go with the second approach of getting and setting the current values of slider in performincrement/performdecrement functions in SliderAccessibility.m. I am not sure if this will be correct approach, but it will be less intrusive and will not need a CSR.

I would suggest to fix it properly by filing CSR for the functional change. The idea of new implementation is not only to make it compatible with the new Apple API, but also to make it cleaner, more consistent and easier to support in the future.

**Comment by pankaj-bansal** [comment]
> If it does not work on Windows then I suggest checking that the new functionality requested via CSR will help it as well.

I could not find any key shortcuts where JAWS will consume the key pressed and will be used to interact with the component like the VO is doing here. This similar functionality is present in JSpinner already and JAWS says similar thing when a Spinner is in focus "To set the value, press the arrow keys or type the value". When user presses the keys, it is consumed by Java and value is incremented/decremented as expected. But the value is not being set by JAWS like VO if used with VO keys. So I am not sure how to verify if the changes will be useful for windows as well. I think if there is a way user can interact with Spinner or Slider, then these changes should be useful and if it is not possible, it will not be useful in both. These changes does make the Slider more consistent with Spinner and easy to maintain and cleaner.
If you happen to have some idea about how can a user interact with Sliders or Spinners using JAWS, please let me know. I will definitely verify if the changes would be helpful for Windows.

**Comment by mrserb** [comment]
> If you happen to have some idea about how can a user interact with Sliders or Spinners using JAWS, please let me know. I will definitely verify if the changes would be helpful for Windows.

How the native applications works under JAWS?

**Comment by pankaj-bansal** [comment]
> > If you happen to have some idea about how can a user interact with Sliders or Spinners using JAWS, please let me know. I will definitely verify if the changes would be helpful for Windows.
> 
> How the native applications works under JAWS?

I tried running a Windows native settings app like "Display Settings" with JAWS. It has a Slider "Brightness". JAWS says same thing when the Slider was brought in focus like "To increase or decrease, use the arrow keys". When I press the arrow keys, the values change. But I don't know if the value changed because the Windows Native dialog handles the keys and changed the value, or JAWS consumed the event and then changed the value. It looks like the Dialog must have handled the keys as there are no special keys mentioned for JAWS to interact with the component like the VO.  I am not sure how I can verify this on Windows native apps. 
I did try to run the JSpinner with JAWS and I see that if Java does not handle the keys, the value in Spinner does not change. This means JAWS does not interact with the component and the accessibility functionality was added for MacOS only. Same thing is being done for JSlider.

File: src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m (hunk 1 above)
Comment (pankaj-bansal) [reply src/java.desktop/macosx/native/libawt_lwawt/awt/a11y/CommonComponentAccessibility.m:59]: Done

**Comment by mrserb** [comment]
> It looks like the Dialog must have handled the keys as there are no special keys mentioned for JAWS to interact with the component like the VO

Did you check this list of hotkeys?
https://www.freedomscientific.com/training/jaws/hotkeys
https://freedomscientific.github.io/VFO-standards-support/aria.html
Something like:
    RIGHT ARROW and UP ARROW increase the value of the slider.
    LEFT ARROW and DOWN ARROW decrease the value of the slider.
    HOME and END move to the minimum and maximum values of the slider.
    PAGE UP and PAGE DOWN increment or decrement the slider by a given amount.

=======================
My main point is that if we will change some shared code we should care about both supported platforms.

**Comment by pankaj-bansal** [comment]
> > It looks like the Dialog must have handled the keys as there are no special keys mentioned for JAWS to interact with the component like the VO
> 
> Did you check this list of hotkeys?
> https://www.freedomscientific.com/training/jaws/hotkeys
> https://freedomscientific.github.io/VFO-standards-support/aria.html
> Something like:
> RIGHT ARROW and UP ARROW increase the value of the slider.
> LEFT ARROW and DOWN ARROW decrease the value of the slider.
> HOME and END move to the minimum and maximum values of the slider.
> PAGE UP and PAGE DOWN increment or decrement the slider by a given amount.
> 
> =======================
> My main point is that if we will change some shared code we should care about both supported platforms.

I tried the hot keys with the both native dialog ("Display Settings" with a "Brightness" Slider ) and java applications. 
Both Java and native dialog  are themselves handling the RIGHT ARROW, UP ARROW, LEFT ARROW, DOWN ARROW, HOME and END buttons for JSlider and user can change slider value using these keys without JAWS also. but they are not handling the PAGE UP and PAGE DOWN button.
PAGE UP and PAGE DOWN button don't work with JAWS either, meaning JAWS is not able to handle the keys and is not able to change component values for both native dialogs and Java application. Also, I am not able to find any component in Swing, where JAWS can change the component values using any hot key. Please correct me if I am wrong in this and JAWS can actually manipulate some component. Also, in the old implementation for MacOS , the VO can not change the Spinner/Slider values and we are adding this functionality now only. 
In addition to this, the functionality we are adding in slider can be used with JAWS also if we decide to provide the support for the same later. Currently, I do not see JAWS being able to change any component value in swing, so it will not be able to change Slider value too.

**Comment by mrserb** [comment]
Ok, please file a similar bug for the windows platform.

**Comment by pankaj-bansal** [comment]
> Ok, please file a similar bug for the windows platform.

I have created the bug for windows here https://bugs.openjdk.java.net/browse/JDK-8264598.
I have also created the CSR for the current issue.

**Comment by pankaj-bansal** [comment]
@mrserb @azuev-java Any other comments on this? Can you please look at the CSR also?

**Comment by mrserb** [comment]
I have no additional comments.

**Review by azuev-java — APPROVED** [review]

**Pull request closed without merging**